packages/useful-types/test-d/types.test-d.ts

```diff
@@ -13,377 +8,288 @@ import type {
+      list?: {nested?: {required?: string | undefined}}[];
+    }>();
+  });
+
+  it('handles readonly arrays', () => {
+    interface ReadOnlyListObj {
+      list: ReadonlyArray<Base>;
+    }
+    expect<DeepPartial<ReadOnlyListObj>>().type.toMatch<{
+      list?: ReadonlyArray<{}> | undefined;
+    }>();
+  });
+
+  it('returns primitives unaltered', () => {
+    expect<DeepPartial<string>>().type.toBeString();
+  });
+
+  it('respects array elements', () => {
+    interface RespectArrayElements {
+      list: (string | number)[] | null;
+    }
+
+    expect<DeepPartial<RespectArrayElements>>().type.toEqual<{
+      list?: (string | number)[] | null;
+    }>();
+  });
+
+  it('handles readonly objects', () => {
+    type PresetOption = Readonly<{
+      name?: string;
+      id: string;
+    }>;
+
+    type Preset = Readonly<{
+      options: ReadonlyArray<PresetOption>;
+      editableOptions: PresetOption[];
+    }>;
+
+    type PresetState = Readonly<{
+      list?: ReadonlyArray<Preset>;
+    }>;
+    type State = Readonly<{
+      preset: PresetState;
+    }>;
+
+    expect<DeepPartial<State>>().type.toMatch<{}>();
+
+    expect<DeepPartial<State>>().type.toMatch<{
+      readonly preset?: {
+        readonly list?: ReadonlyArray<{
+          readonly options?: ReadonlyArray<{readonly name?: string}>;
+        }>;
+      };
+    }>();
+
+    expect<DeepPartial<State>>().type.toMatch<{
+      readonly preset?: {} | undefined;
+    }>();
+
+    expect<DeepPartial<State>>().type.toMatch<{
+      readonly preset?: {
+        readonly list?: ReadonlyArray<{}> | undefined;
+      };
+    }>();
+
+    expect<DeepPartial<State>>().type.toMatch<{
+      readonly preset?: {
+        readonly list?: ReadonlyArray<{
+          readonly options?: ReadonlyArray<{}> | undefined;
+        }>;
+      };
+    }>();
+
+    expect<DeepPartial<State>>().type.toMatch<{
+      readonly preset?: {
+        readonly list?: ReadonlyArray<{
+          readonly editableOptions?: {}[] | undefined;
+        }>;
+      };
+    }>();
+  });
 });
 
-expectNotAssignable<DeepPartial<State>>({
-  preset: {
-    list: [undefined],
-  },
-});
+describe('IfEmptyObject', () => {
+  it('checks if an object is empty', () => {
+    expect<IfEmptyObject<{}, true>>().type.toEqual<true>();
+    expect<IfEmptyObject<{foo: string}, never, false>>().type.toEqual<false>();
 
-expectNotAssignable<DeepPartial<State>>({
-  preset: {
-    list: [{options: [undefined]}],
-  },
-});
-expectNotAssignable<DeepPartial<State>>({
-  preset: {
-    list: [{editableOptions: [undefined]}],
-  },
+    expect<IfEmptyObject<{foo: string}, true>>().type.not.toEqual<true>();
+    expect<IfEmptyObject<boolean, true>>().type.not.toEqual<true>();
+  });
 });
 
-/**
- * should not be able to add undefined, need to fix this
- */
-expectNotAssignable<DeepPartial<RespectArrayElements>>({
-  // this should error, undefined is not a valid type for list
-  list: [undefined, 'stable', 2],
```

Comment:
Hard to say what was the problem and why "need to fix this" note is left here. As far as I can get, it all works as expected.